### PR TITLE
made interval generic on the JVM side

### PR DESCRIPTION
**Pull request by cseed**
Previous patches made the interval _type_ generic (e.g. Interval[T] instead of just Interval which implied interval of locus).  This patch makes the intervals themselves generic.  The interval doesn't carry the type (it's just a container for the start, end) so various interval operations need the ordering on the type that the interval is over to be passed in.  Then everything else is follows from this.

The next PR makes the intervals generic on the Python side.

Along the way, I allows intervals to be improper, e.g., [7, 5), which are effectively empty.  Note, intervals compare for equality if the endpoints are equal, so [7, 5) != [8, 5) even though they are both empty.

### CODE REVIEW

File: src/main/scala/is/hail/variant/GenomeReference.scala [hunk 1]
```diff
@@ -17,6 +17,7 @@ import scala.language.implicitConversions
 abstract class GRBase extends Serializable {
   val variant: TVariant
   val locus: TLocus
+  val intervalType: TInterval
```
Comment (jigold) [inline src/main/scala/is/hail/variant/GenomeReference.scala]: Can you make it `variantType` and `locusType` to be consistent?

File: src/main/scala/is/hail/variant/Locus.scala [hunk 2]
```diff
@@ -53,26 +50,26 @@ object Locus {
...
+      parseAll[Interval](locus_interval, input) match {
         case Success(r, _) => r
         case NoSuccess(msg, next) => fatal(s"invalid interval expression: `$input': $msg")
       }
     }
 
-    def interval(gr: GRBase): Parser[Interval[Locus]] = {
-      implicit val locusOrd = gr.locusOrdering
+    // FIXME make end correct
+    def locus_interval: Parser[Interval] = {
```
Comment (jigold) [inline src/main/scala/is/hail/variant/Locus.scala]: This is fine for now, but my PR will need the GRBase. I check the Locus has valid contig and pos. I also fixed the end in my branch.

File: src/main/scala/is/hail/utils/IntervalTree.scala [hunk 3]
```diff
@@ -1,99 +1,90 @@
...
 import scala.collection.mutable
 import scala.language.implicitConversions
-import scala.math.Ordering.Implicits._
 import scala.reflect.ClassTag
 
 // interval inclusive of start, exclusive of end: [start, end)
-case class Interval[T](start: T, end: T)(implicit ev: Ordering[T]) extends Serializable {
-  import ev._
+case class Interval(start: Any, end: Any) extends Serializable {
+  def contains(pord: ExtendedOrdering, position: Any): Boolean = pord.gteq(position, start) && pord.lt(position, end)
```
Comment (jigold) [inline src/main/scala/is/hail/utils/IntervalTree.scala]: What happens if you create an Interval where start and end are two different types? Is that tested for anywhere?

Comment (cseed) [reply src/main/scala/is/hail/utils/IntervalTree.scala]: That's not allowed, the same way creating an `IndexedSeq[Any]("foo", 5)` is not allowed to represent a value of type `Array[Int32]`.  It is tested by TInterval._typeCheck, which verifies the start and end both are instances of `pointType`.

```
  def _typeCheck(a: Any): Boolean = a.isInstanceOf[Interval] && {
    val i = a.asInstanceOf[Interval]
    pointType.typeCheck(i.start) && pointType.typeCheck(i.end)
  }
```

File: src/main/scala/is/hail/variant/Locus.scala (hunk 2 above)
Comment (cseed) [reply src/main/scala/is/hail/variant/Locus.scala]: OK, great.  Or if your PR is ready to go in, I'm happy to let it go in first and rebase.

Comment (jigold) [reply src/main/scala/is/hail/variant/Locus.scala]: Your call. @danking needs to review it again #2558.

**Pull request merged**